### ISSUE TITLE
recipe for target 'pin_exec' failed

### ISSUE THREAD

**Issue by Upasna55** [hpsresearchgroup/scarab#116]
Asking a question on the Scarab Issues page allows the entire Scarab community to benefit from the discussion. It also decreases response time as any of the Scarab administrators can answer these questions. 

Please type your questions here!

**Comment by Upasna55** [comment]
Hello, when I execute **make** in **scarab/src** directory, I get the following error:
![Screenshot from 2021-11-05 17-59-05](https://user-images.githubusercontent.com/91068864/140510978-f1200ddc-57fe-42e1-8f03-5e13c4af3c70.png)
Can you please help me out to resolve this error?

**Comment by spruett** [comment]
It looks like you have not installed PIN. 

Please see the system requirements page for instructions on how to download PIN: 
https://github.com/hpsresearchgroup/scarab/blob/master/docs/system_requirements.md

Once installed, create a PIN_ROOT environment variable that points to the directory where you installed PIN. 

After that, run make in the src directory again and it should work.

**Comment by Upasna55** [comment]
I have already installed pin-3.15 and pinplay-3.5 in the **src** directory and set the PIN_ROOT path correctly.

**Comment by spruett** [comment]
Have you verified that the path mentioned in the error message exists? (`/home/tvkalyan/DATA/upasna/scarab/src/pin-3.5/extras/stlport/include/algorithm`)

**Comment by Upasna55** [comment]
Stephen, /home/tvkalyan/DATA/upasna/scarab/src/pin-3.5/extras/stlport/include/algorithm path exists.

PS: I havn't renamed  pin-3.15-98253-gb56e429b1-gcc-linux to pin-3.15. Similarly pinplay-3.5 is pinplay-drdebug-3.5-pin-3.5-97503-gac534ca30-gcc-linux. And, there is one more **pin** directory already available in the simulator setup.

**Comment by spruett** [comment]
Hmmm... I think I need some clarification on your last message. 

Let me repeat what I am understanding, then you correct me where appropriate. 

You are saying that you downloaded a directory called `pin-3.15-98253-gb56e429b1-gcc-linux` and renamed the director to `pin-3.15` AND you downloaded a director called `pinplay-drdebug-3.5-pin-3.5-97503-gac534ca30-gcc-linux` and renamed it to `pinplay-3.5`. 

You also say, "And, there is one more pin directory already available in the simulator setup." and I am not quite sure what you mean by this. 

But lets talk about the `pin-3.15` and `pinplay-3.5` directories first. 

First, it does not matter that you renamed the directories, that does not effect anything.  

Second, you do not need to install both `pin-3.15` AND `pinplay-3.5`. You only need one of them to run Scarab. `pinplay-3.5` is required if you want to use our checkpoint infrastructure, however some users have had problems getting that older version of pin running on newer versions of Ubuntu. So for that reason, you may need pin-3.15. However that compatibility issue is NOT the compile error you are seeing here. This compile error is being caused because the makefile cannot find the specified path (`/home/tvkalyan/DATA/upasna/scarab/src/pin-3.5/extras/stlport/include/algorithm`). So to be clear, it does not matter that you installed both pin and pinplay, but you only need one of them to run Scarab, and you should set `PIN_ROOT` to point to the one that you wish Scarab to use. 

Third, your error message seems to indicate that you set `$PIN_ROOT` to `/home/tvkalyan/DATA/upasna/scarab/src/pin-3.5`, which means it is pointing to a directory called `pin-3.5` and NOT one of the two directories that you said you downloaded (`pin-3.15` and `pinplay-3.5`). It is possible this is what is causing the error.

Can you run the following commands and paste the output of each here so I can further understand your problem:
```
> echo $PIN_ROOT
> ls $PIN_ROOT/extras/stlport/include/algorithm
```

**Comment by Upasna55** [comment]
Just to correct, I wrote that I have not renamed pin. 
**echo $PIN_ROOT** outputs /home/tvkalyan/DATA/upasna/scarab/src/pinplay-drdebug-3.5-pin-3.5-97503-gac534ca30-gcc-linux
**ls $PIN_ROOT/extras/stlport/include/algorithm** outputs /home/tvkalyan/DATA/upasna/scarab/src/pinplay-drdebug-3.5-pin-3.5-97503-gac534ca30-gcc-linux/extras/stlport/include/algorithm

**Comment by spruett** [comment]
But the path `/home/tvkalyan/DATA/upasna/scarab/src/pin-3.5/extras/stlport/include/algorithm` also exists?

Based on your messages, it seems like the `pin_exec` Makefile is looking for PIN at this path: `/home/tvkalyan/DATA/upasna/scarab/src/pin-3.5`

When it should be looking for PIN at this path: `/home/tvkalyan/DATA/upasna/scarab/src/pinplay-drdebug-3.5-pin-3.5-97503-gac534ca30-gcc-linux`

Strangely, this makefile is simply the makefile that is distributed with PIN. So as long as everything in the pin directory that you downloaded is still in tact and the `PIN_ROOT` environment variable is set correctly, then this makefile should work. 

I am going to have you run a few more commands for me. 

First double-check that PIN_ROOT is set to `/home/tvkalyan/DATA/upasna/scarab/src/pinplay-drdebug-3.5-pin-3.5-97503-gac534ca30-gcc-linux` and confirm that this file exists: `$PIN_ROOT/extras/stlport/include/algorithm`

Now, run the following commands and paste the output here:
```
> cd src/pin/pin_exec
> make 
```

**Comment by Upasna55** [comment]
Stephen, I got the following outputs:

![Screenshot from 2021-11-06 10-58-50](https://user-images.githubusercontent.com/91068864/140599160-4d718b28-97d9-4e04-9558-0fd1beaa060a.png)
:
![Screenshot from 2021-11-06 10-59-03](https://user-images.githubusercontent.com/91068864/140599163-e9a673c6-6d01-4363-8828-bfa1b0e39177.png)

**Comment by spruett** [comment]
Ok, it looks like the initial problem has gone away. 

### What was going on:

I think originally you did not have your `PIN_ROOT` environment variable pointing to the correct place. The original log that you posted indicates that your `PIN_ROOT` was set to `/home/tvkalyan/DATA/upasna/scarab/src/pin-3.5`, which was not correct. At some point, it seems like you switched the `PIN_ROOT` over to `/home/tvkalyan/DATA/upasna/scarab/src/pinplay-drdebug-3.5-pin-3.5-97503-gac534ca30-gcc-linux`, which is correct. We were able to confirm this by running `ls $PIN_ROOT/extras/stlport/include/algorithm`, which confirmed that the file in question (the file that was originally missing when you first posted) now exists. 

### What is going on now:

Now, running `make` in the `src/pin/pin_exec` directory completely successfully. This was the part of the build that was originally failing, so clearly we have fixed the original problem. 

Now, there is a new problem. As the error message now states:

> Could not find python3 in the system

So I will again refer you to our [System Requirements](https://github.com/hpsresearchgroup/scarab/blob/master/docs/system_requirements.md). Here you will see that we require Python3 as well as several other tools. Please be sure you have installed the necessary system requirements and that these programs are visible on your `$PATH`.

**Comment by Upasna55** [comment]
Yes, Stephen, the problem was in the PIN path. From the system requirements page, I installed the required packages. Following is the packages' version that I have on my system:
![Screenshot from 2021-11-07 11-33-00](https://user-images.githubusercontent.com/91068864/140634456-6971afd4-ee7a-4f40-9c95-97df341bb5e7.png)

Even though I have Python3, still the error in `scarab/src/` make is due to Python3

![Screenshot from 2021-11-07 11-33-24](https://user-images.githubusercontent.com/91068864/140634508-4ec9dc18-a8dd-41e4-b98a-66679dd42a09.png)

**Comment by spruett** [comment]
Does `python3` exist?

try running:
```
python3 —version 
```

**Comment by Upasna55** [comment]
Stephen, the issue is resolved now. Thanks a lot.

**Comment by Upasna55** [comment]
Stephen, when I execute `python scarab_test_qsort.py tmp` in `scarab/utils/qsort`, I get yet another error due to PIN as follows:

`failed to execute /misc/pool1/pin/pinplay-drdebug-3.5-pin-3.5-97503-gac534ca30-gcc-linux/pin: No such file or directory`

![Screenshot from 2021-11-08 20-42-50](https://user-images.githubusercontent.com/91068864/140768179-89335cb4-3ca6-4477-8c8b-f65a737259e1.png)

**Comment by spruett** [comment]
This error is happening because `PIN_ROOT` is not set. 

The script `scarab_test_qsort.py` calls the script `scarab_launch.py` to run Scarab. `scarab_launch.py` uses the paths defined in [scarab_paths.py](https://github.com/hpsresearchgroup/scarab/blob/master/bin/scarab_globals/scarab_paths.py) to find PIN. 

By default [scarab_paths.py](https://github.com/hpsresearchgroup/scarab/blob/master/bin/scarab_globals/scarab_paths.py) checks the environment variable `PIN_ROOT` to find pin. If the environment variable is not set, then it uses the default path specified in  [scarab_paths.py](https://github.com/hpsresearchgroup/scarab/blob/master/bin/scarab_globals/scarab_paths.py). As a placeholder, that path is set to `/misc/pool1/pin/pinplay-drdebug-3.5-pin-3.5-97503-gac534ca30-gcc-linux`, which does not exist on your machines. Hence the error message. 

So in short, to solve this problem please be sure `PIN_ROOT` is set properly, just as was needed for compile. You can setup `PIN_ROOT` in your `.profile` so that it is setup for you automatically each time you log into a machine.

**Comment by Upasna55** [comment]
Thank you, Stephen. Can you please help me to resolve `Connection reset by peer` error? This time, the error shouldn't be because of PIN_ROOT as I have permanently set this environment variable value. I get the following output when I execute `python scarab_test_qsort.py tmp` in `scarab/utils/qsort/`.
![Screenshot from 2021-11-09 10-43-08](https://user-images.githubusercontent.com/91068864/140867753-c653fb94-aca0-4157-9d4f-ec49ce974351.png)

**Comment by spruett** [comment]
It seems like PIN is not running properly. Scarab is essentially just a glorified pintool, so if PIN is not functioning, then Scarab will not work either. 

I would try running one of the pintools that they distribute with PIN to make sure everything on that end is functioning properly. Once you are confident PIN is working, come back to Scarab and try running it again. 

Here are some directions on how to run the example pintools: https://software.intel.com/sites/landingpage/pintool/docs/71313/Pin/html/

**Comment by Upasna55** [comment]
Ok, Stephen. Thanks.

**Comment by spruett** [comment]
Closing this issue due to inactivity. Feel free to reopen if needed.
